### PR TITLE
rewrite hash log

### PR DESCRIPTION
**Pull request by fbenkstein**

### CODE REVIEW

File: src/build_log.cc [hunk 1]
```diff
@@ -6,7 +6,7 @@
 //
 //     http://www.apache.org/licenses/LICENSE-2.0
 //
-// Unless required by applicable law or agreed to in writing, software
+j/ Unless required by applicable law or agreed to in writing, software
```
Comment (fbenkstein) [inline src/build_log.cc:9]: Oops.

File: src/disk_interface.h [hunk 2]
```diff
@@ -19,8 +19,15 @@
...
 
+#ifdef _WIN32
+#include "win32port.h"
+#else
+#include <stdint.h>
+#endif
+
 #include "timestamp.h"
 
+
```
Comment (fbenkstein) [inline src/disk_interface.h:30]: Delete.

File: src/disk_interface.h [hunk 3]
```diff
@@ -39,11 +46,22 @@ struct FileReader {
                           string* err) = 0;
 };
 
+/// Interface for hashing files from disk.
+struct FileHasher {
+  typedef uint64_t Hash;
```
Comment (fbenkstein) [inline src/disk_interface.h:51]: 32 bit.

File: src/hash_log.cc [hunk 4]
```diff
@@ -23,484 +23,388 @@
 #include <unistd.h>
 #endif
 
+#include <iostream>
+#include <algorithm>
```
Comment (fbenkstein) [inline src/hash_log.cc:27]: Unnecessary.

File: src/hash_log.cc [hunk 5]
```diff
@@ -23,484 +23,388 @@
...
-    ++total_values_;
   }
-  return true;
-}
 
-bool HashLog::UpdateHash(Node* node, const hash_variant variant,
-                         string* err, const bool force, hash_t* result) {
-  // initialize the result
-  if (result) { *result = 0; }
+  delete entry;
```
Comment (fbenkstein) [inline src/hash_log.cc:156]: next_id_ = *ids.rbegin() + 1;

File: src/hash_log.cc [hunk 6]
```diff
@@ -23,484 +23,388 @@
...
+    if (file_ != NULL && !WriteEntry(node, it->second, err))
+      return false;
   }
-  return false;
+
+  Hash new_hash;
+
+  if (is_input) {
+    new_hash = it->second->input_hash_;
+    *acc ^= it->second->input_hash_;
```
Comment (fbenkstein) [inline src/hash_log.cc:322]: Find something before than XOR.

File: src/hash_log.h [hunk 7]
```diff
@@ -1,4 +1,5 @@
-// Copyright 2014 Matthias Maennich (matthias@maennich.net).
+// Copyright 2014 Matthias Maennich (matthias@maennich.net)
+//           2016 SAP SE
 // All Rights Reserved.
```
Comment (fbenkstein) [inline src/hash_log.h:3]: Update in hash_log.cc as well.

File: src/ninja.cc [hunk 8]
```diff
@@ -903,6 +911,39 @@ bool NinjaMain::OpenDepsLog(bool recompact_only) {
...
+  if (recompact_only) {
+    bool success = hash_log_.Recompact(path, &err);
+    if (!success)
+      Error("failed recompaction: %s", err.c_str());
+    return success;
+  }
+
+  if (!config_.dry_run) {
+    if (!hash_log_.OpenForWrite(path, &err)) {
+      Error("opening deps log: %s", err.c_str());
```
Comment (metti) [inline src/ninja.cc:939]: hash log

File: src/hash_log.cc [hunk 9]
```diff
@@ -23,484 +23,388 @@
...
-  if (it == hash_map_.end() || it->first         != key
-                            || it->second.hash_  != hash
-                            || it->second.mtime_ != mtime) {
+  bool valid_header = true;
+  int version = 0;
+  if (!fgets(buf, sizeof(buf), f) || fread(&version, 4, 1, f) < 1)
+    valid_header = false;
+  if (!valid_header || strcmp(buf, kFileSignature) != 0 ||
+      version != kCurrentVersion) {
+    if (version == 1)
```
Comment (metti) [inline src/hash_log.cc:71]: why the version==1 special case?

File: src/hash_log.cc [hunk 10]
```diff
@@ -23,484 +23,388 @@
...
+      *err = "hash log version change; rebuilding";
+    else
+      *err = "bad hash log signature or version; starting over";
+    fclose(f);
+    unlink(path.c_str());
+    // Don't report this as a failure.  An empty deps log will cause
```
Comment (metti) [inline src/hash_log.cc:77]: hash log

File: src/hash_log.cc [hunk 11]
```diff
@@ -23,484 +23,388 @@
...
+  // Rebuild the log if there are too many dead records.
+  size_t kMinCompactionEntryCount = 1000;
+  size_t kCompactionRatio = 3;
+  if (total_entry_count > kMinCompactionEntryCount &&
+      total_entry_count > entries_.size() * kCompactionRatio) {
+    needs_recompaction_ = true;
   }
 
-  // check whether we need to update the hash
+  // for (Entries::iterator it = entries_.begin(); it != entries_.end(); ++it)
```
Comment (metti) [inline src/hash_log.cc:187]: remove

**Pull request closed without merging**